protocol/network/4-architecture.md

```diff
@@ -0,0 +1,126 @@
+4 Architecture
+==============
+
+libp2p was designed around the Unix Philosophy, creating smaller components, easier to understand and to test. These components should also be able to be swapped in order to accomodate different technologies or scenarios and also make it that it is upgradable over time.
+
+Although different Peers can support different protocols depending on their capabilities, any Peer can act as a dialer and/or a listener for connections from other Peers, connections that once established can be reused from both ends, removing the distinction between clients and servers.
+
+libp2p interface acts as a thin veneer to three subsystems that are required in order for peers to be able to communicate. These subsystems are allowed to be built on top of other subsystems as long as they respect the standardized interface. The main 3 subsystems are:
+
+- Peer Routing - Mechanism to find a Peer in a network. This Routing can be done recursively, iteratively or even in a broadcast/multicast mode.
+- Swarm - Handles everything that touches the 'opening a stream' part of libp2p, from protocol muxing, stream muxing, NAT Traversal, Connection Relaying, while being multitransport
```

Comment:
@diasdavid i think you merged "NAT Traversal" "Connection Relaying" and "Protocol Muxing" into your node swarm, but i do not think think this should be merged in the libp2p spec, nor on Go. The go swarm only handles transport multiplexing (connections/streams) to many peers. NAT and Relay are separate protocols. In Go, the "Host" handles the protocol multiplexing, and composes (uses) the NAT and Relay protocols.

So, this should not be "tree main subsystems", but rather "multiple subsystems" brought together under the `libp2p.Node` (I think "Node" is a better word than "Host" or "Peer").